"panic: city manager did not find the city"

### Graphical example
![image](https://github.com/user-attachments/assets/4b82f664-aed7-48fe-bf5f-73ea9e615eb7)

Meeple placed on (+1, +3) field

### Regression test
https://github.com/YetAnotherSpieskowcy/Carcassonne-Engine/commit/1a9ea464b9f4183857f475af7555bc29cdefad54

### Test output
```
$ go test -count=1 ./pkg/game -run TestRegressionGHIssue71
--- FAIL: TestRegressionGHIssue71 (0.00s)
panic: city manager did not find city: elements.PlacedFeature{Feature:feature.Feature{FeatureType:2, ModifierType:0, Sides:0x3c}, Meeple:elements.Meeple{Type:0x0, PlayerID:0x0}} at position position.Position{x:1, y:2} [recovered]
	panic: city manager did not find city: elements.PlacedFeature{Feature:feature.Feature{FeatureType:2, ModifierType:0, Sides:0x3c}, Meeple:elements.Meeple{Type:0x0, PlayerID:0x0}} at position position.Position{x:1, y:2}

goroutine 6 [running]:
testing.tRunner.func1.2({0x553d80, 0xc000032420})
	/usr/lib/go-1.22/src/testing/testing.go:1631 +0x24a
testing.tRunner.func1()
	/usr/lib/go-1.22/src/testing/testing.go:1634 +0x377
panic({0x553d80?, 0xc000032420?})
	/usr/lib/go-1.22/src/runtime/panic.go:770 +0x132
github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/game/field.(*Field).Expand(0xc0000b2da8, {0x5bfdc8, 0xc0000d2080}, {{0xc0000a91a0?, 0x47?, 0x4a?}})
	/home/ubuntu/work/Carcassonne-Engine/pkg/game/field/field.go:109 +0x7c5
github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/game.TestRegressionGHIssue71(0xc0000c84e0)
	/home/ubuntu/work/Carcassonne-Engine/pkg/game/field_test.go:240 +0xf7d
testing.tRunner(0xc0000c84e0, 0x590138)
	/usr/lib/go-1.22/src/testing/testing.go:1689 +0xfb
created by testing.(*T).Run in goroutine 1
	/usr/lib/go-1.22/src/testing/testing.go:1742 +0x390
FAIL	github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/game	0.003s
FAIL
```

--- a/pkg/game/city/city_manager.go
+++ b/pkg/game/city/city_manager.go
@@ -84,7 +84,16 @@ func (manager Manager) findCitiesToJoin(foundCities map[side.Side]int, tile elem
 			if sides.HasSide(mask) {
 				c, ok := foundCities[mask]
 				if ok {
-					cToJoin = append(cToJoin, c)
+					notConsidered := true
+					for _, val := range cToJoin {
+						if val == c {
+							notConsidered = false
+							break
+						}
+					}
+					if notConsidered {
+						cToJoin = append(cToJoin, c)
+					}
 				}
 			}
 			mask = mask.Rotate(1)

--- a/pkg/game/city/city_manager.go
+++ b/pkg/game/city/city_manager.go
@@ -77,28 +77,26 @@ func (manager Manager) findCitiesToJoin(foundCities map[side.Side]int, tile elem
 	citiesToJoin := map[elements.PlacedFeature][]int{}
 	cityFeatures := tile.GetFeaturesOfType(feature.City)
 	for _, cityFeature := range cityFeatures {
-		var cToJoin = []int{}
+		var cityIndexesToJoin = []int{}
 		sides := cityFeature.Feature.Sides
-		mask := side.Top
-		for range 4 {
+		for _, mask := range side.PrimarySides {
 			if sides.HasSide(mask) {
-				c, ok := foundCities[mask]
+				neighbourCityIndex, ok := foundCities[mask]
 				if ok {
-					notConsidered := true
-					for _, val := range cToJoin {
-						if val == c {
-							notConsidered = false
+					alreadyChecked := false
+					for _, cityIndex := range cityIndexesToJoin {
+						if cityIndex == neighbourCityIndex {
+							alreadyChecked = true
 							break
 						}
 					}
-					if notConsidered {
-						cToJoin = append(cToJoin, c)
+					if !alreadyChecked {
+						cityIndexesToJoin = append(cityIndexesToJoin, neighbourCityIndex)
 					}
 				}
 			}
-			mask = mask.Rotate(1)
 		}
-		citiesToJoin[cityFeature] = cToJoin
+		citiesToJoin[cityFeature] = cityIndexesToJoin
 	}
 	return citiesToJoin
 }
@@ -111,19 +109,19 @@ func (manager *Manager) UpdateCities(tile elements.PlacedTile) {
 		citiesToRemove := make([]int, 0)
 		citiesToJoin := manager.findCitiesToJoin(foundCities, tile)
 		// join cities
-		for f, cToJoin := range citiesToJoin {
-			if len(cToJoin) == 0 {
+		for f, cityIndexesToJoin := range citiesToJoin {
+			if len(cityIndexesToJoin) == 0 {
 				toAppend := []elements.PlacedFeature{f}
 				manager.cities = append(manager.cities, NewCity(tile.Position, toAppend))
 			} else {
-				if len(cToJoin) > 1 {
-					for _, cityIndex := range cToJoin[1:] {
-						manager.cities[cToJoin[0]].JoinCities(manager.cities[cityIndex])
+				if len(cityIndexesToJoin) > 1 {
+					for _, cityIndex := range cityIndexesToJoin[1:] {
+						manager.cities[cityIndexesToJoin[0]].JoinCities(manager.cities[cityIndex])
 						citiesToRemove = append(citiesToRemove, cityIndex)
 					}
 				}
 				toAdd := []elements.PlacedFeature{f}
-				manager.cities[cToJoin[0]].AddTile(tile.Position, toAdd)
+				manager.cities[cityIndexesToJoin[0]].AddTile(tile.Position, toAdd)
 			}
 		}
 		// remove cities that were merged into another city

--- a/pkg/game/city/city_manager_test.go
+++ b/pkg/game/city/city_manager_test.go
@@ -120,6 +120,68 @@ func TestJoinCitiesOnAddCityNotClosed(t *testing.T) {
 	}
 }
 
+func TestJoinCitiesFourEdgeCity(t *testing.T) {
+	a1 := elements.ToPlacedTile(tiletemplates.TwoCityEdgesCornerConnected())
+	a1.Position = position.New(1, 1)
+	manager := NewCityManager()
+	manager.UpdateCities(a1)
+
+	a2 := elements.ToPlacedTile(tiletemplates.TwoCityEdgesCornerConnectedShield().Rotate(1))
+	a2.Position = position.New(1, 2)
+	manager.UpdateCities(a2)
+
+	a3 := elements.ToPlacedTile(tiletemplates.ThreeCityEdgesConnectedShield())
+	a3.Position = position.New(2, 1)
+	manager.UpdateCities(a3)
+
+	d := elements.ToPlacedTile(tiletemplates.ThreeCityEdgesConnected())
+	d.Position = position.New(1, 3)
+	manager.UpdateCities(d)
+
+	e := elements.ToPlacedTile(tiletemplates.FourCityEdgesConnectedShield())
+	e.Position = position.New(2, 2)
+	manager.UpdateCities(e)
+
+	if len(manager.cities) != 2 {
+		t.Fatalf("expected %#v, got %#v instead", 2, len(manager.cities))
+	}
+}
+
+func TestJoinCitiesFourEdgeCityTwoCitiesConnected(t *testing.T) {
+	a1 := elements.ToPlacedTile(tiletemplates.TwoCityEdgesCornerConnected())
+	a1.Position = position.New(1, 1)
+	manager := NewCityManager()
+	manager.UpdateCities(a1)
+
+	a2 := elements.ToPlacedTile(tiletemplates.TwoCityEdgesCornerConnectedShield().Rotate(1))
+	a2.Position = position.New(1, 2)
+	manager.UpdateCities(a2)
+
+	a3 := elements.ToPlacedTile(tiletemplates.ThreeCityEdgesConnectedShield())
+	a3.Position = position.New(2, 1)
+	manager.UpdateCities(a3)
+
+	b1 := elements.ToPlacedTile(tiletemplates.TwoCityEdgesCornerConnected().Rotate(1))
+	b1.Position = position.New(2, 3)
+	manager.UpdateCities(b1)
+
+	b2 := elements.ToPlacedTile(tiletemplates.TwoCityEdgesCornerConnected().Rotate(2))
+	b2.Position = position.New(3, 3)
+	manager.UpdateCities(b2)
+
+	b3 := elements.ToPlacedTile(tiletemplates.TwoCityEdgesCornerConnected().Rotate(3))
+	b3.Position = position.New(3, 2)
+	manager.UpdateCities(b3)
+
+	e := elements.ToPlacedTile(tiletemplates.FourCityEdgesConnectedShield())
+	e.Position = position.New(2, 2)
+	manager.UpdateCities(e)
+
+	if len(manager.cities) != 1 {
+		t.Fatalf("expected %#v, got %#v instead", 1, len(manager.cities))
+	}
+}
+
 func TestForceScore(t *testing.T) {
 	var expectedScore uint32 = 1
 	var expectedMeepleType elements.MeepleType = elements.NormalMeeple

